### ISSUE TITLE
Dynamic Tag Name Props in React (with TypeScript)

### ISSUE THREAD

**Issue by AleksandrHovhannisyan** [AleksandrHovhannisyan/aleksandrhovhannisyan.com#62]

**Comment by equinusocio** [comment]
Thank you for this tut! Unfortunately this solution doesn't cover the type validation of attributes.
If i set the `tagName` as `"button"` and i add the `type="button"` attribute it doesn't recognise it and throw a type validation error

**Comment by AleksandrHovhannisyan** [comment]
Yeah, unfortunately, that won't work as far as I can tell. You'll need to use a slot or render prop if the tag is being used for more than just decorative/wrapping purposes and needs additional props/attributes. That way, the consumer of your component can customize the rendered element as needed.

**Comment by AleksandrHovhannisyan** [comment]
@lianulloa Yup! Like I mentioned above, if the tag name needs to have any props passed to it, then a tag name won't work (because your click event's typing depends on the underlying element type, and that's dynamic—unless you do a type assertion for every possibility, which you should never try to do). What you're looking for is a `ReactNode` slot in this case. Then, you can easily pass along an `onClick` handler and just render the slot in your component.

**Comment by surjithctly** [comment]
If I'm doing a `<Heading>` component, how do I restrict to use `h1, h2 ... h6` only? instead of every html tag?

**Comment by AleksandrHovhannisyan** [comment]
I've had to do that before too. You have two options: 

1. The component accepts a `level` prop of `1 | 2 | 3 | ... | 6`.
2. The component accepts a string prop of `'h1' | 'h2' | 'h3' | ... | 'h6'`.

If you go with option 1, you can interpolate the number in a string and assert the type as an element. Same for option 2.

**Comment by lexuantien** [comment]
[Microsoft FluentUI support dynamic Tag](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-utilities/src/compose/types.ts#L97). It sounds easy to config but diffcult to get props

**Comment by lexuantien** [comment]
Another dynamic tag come from [react.js beta page](https://github.com/reactjs/reactjs.org/blob/main/beta/src/utils/forwardRefWithAs.tsx#L33). It used to style Heading in MDX page. Because heading can be ` h1  h2 h3 h4 ` tag. The drawback is it doesn't runtime error check like Fluentui.

**Comment by duongductrong** [comment]
In this case, I want to pass a Component to "as" props and by some way it can detect and autocomplete props of this component. What do I have to do to be like that?

Example:
```js
import Button from "...."

function Parent() {
     return <AnotherComponentNotHaveButtonProps as={Button} propsColorOfButton="white" />
}

````

**Comment by AleksandrHovhannisyan** [comment]
@duongductrong Why not just pass the component with its props directly as a slot?

```jsx
import Button from "...."

function Parent() {
     return <Component slotName={<Button propsColorOfButton="white" />} />
}
```

**Comment by duongductrong** [comment]
@AleksandrHovhannisyan I want `Component` extends props of another ReactNode component. Because I use nextjs and default we have <Link /> component for navigation between the pages. So, if the `Component` can be extends any component props, then i can use their props in legacy form.

```jsx
import Link from 'next/link'

export interface ButtonProps {

}

const Button : FC<ButtonProps> = ({ as: Tag = "button", children, ...props }) => {
    return <Tag {...props}>{children}</Tag> 
}

const App = ({ ...props }) => {
    return <Button as={Link} href="your-page">To another page</Button>
}
```
